File: gof-design-patterns/srcs/P01_Singleton/Singleton.java
package P01_Singleton;

public class Singleton {
    // 唯一のインスタンスを保持するstatic変数
    private static Singleton instance;
    
    // privateコンストラクタで外部からのインスタンス化を防止
    private Singleton() {
        // 初期化処理
    }
    
    // インスタンスへのアクセスを提供するpublic staticメソッド
    public static Singleton getInstance() {
        if (instance == null) {
            instance = new Singleton();
        }
        return instance;
    }
    
    // その他のメソッド
    public void doSomething() {
        System.out.println("シングルトンのメソッドが呼ばれました");
    }
}

File: gof-design-patterns/cs/P01_Singleton/Singleton.c
#include <stdio.h>
#include <stddef.h>
#include "Singleton.h"

// doSomethingメソッドの実装
static void singleton_doSomething(Singleton* self) {
    if (self != NULL) {
        printf("シングルトンのメソッドが呼ばれました\n");
    }
}

// 唯一のインスタンスを保持するstatic変数（Javaのprivate staticに相当）
// mallocを使わず、静的に確保
static Singleton instance;

// 初期化フラグ（初回のみ初期化を行うため）
static int initialized = 0;

// インスタンスへのアクセスを提供する関数（JavaのgetInstance()に相当）
Singleton* Singleton_getInstance(void) {
    // 初回呼び出し時のみ初期化
    if (!initialized) {
        // 関数ポインタを設定
        instance.doSomething = singleton_doSomething;
        
        // その他の初期化処理（必要に応じて）
        // instance.someData = 0;
        
        initialized = 1;
    }
    return &instance;
}
